--- freetds-0.91/src/apps/bsqldb.c~	2011-03-13 21:32:46.000000000 +0000
+++ freetds-0.91/src/apps/bsqldb.c	2014-12-28 18:04:14.625622727 +0000
@@ -545,7 +545,7 @@
 							perror("could not write to output file");
 							exit(EXIT_FAILURE);
 						}
-						fprintf(stdout, metadata[c].format_string); /* col/row separator */
+						fputs(metadata[c].format_string, stdout); /* col/row separator */
 						continue;
 					}
 					switch (data[c].status) { /* handle nulls */
