--- freetds-0.95.8/configure.ac.orig	2015-06-22 11:00:49.000000000 +0200
+++ freetds-0.95.8/configure.ac	2015-06-25 21:14:01.183204149 +0200
@@ -568,12 +568,10 @@
 	if test -f "$with_unixodbc/bin/odbc_config"; then
 		ODBC_CONFIG="$with_unixodbc/bin/odbc_config"
 		ODBC_INC=`$ODBC_CONFIG --include-prefix`
-		LDFLAGS="$LDFLAGS -L`$ODBC_CONFIG --lib-prefix`"
 	else
 		# if not available do not use system default
 		ODBC_CONFIG=""
 		ODBC_INC="$with_unixodbc/include"
-		LDFLAGS="$LDFLAGS -L$with_unixodbc/lib"
 	fi
 	test -r "$ODBC_INC/sql.h" || AC_ERROR([sql.h not found])
 	test -r "$ODBC_INC/sqlext.h" || AC_ERROR([sqlext.h not found])
--- freetds-0.95.8/m4/check_openssl.m4.orig	2015-05-05 19:54:59.000000000 +0200
+++ freetds-0.95.8/m4/check_openssl.m4	2015-06-25 21:13:31.026538747 +0200
@@ -18,7 +18,6 @@
     else
         echo "OpenSSL found in $ssldir"
         NETWORK_LIBS="$NETWORK_LIBS -lssl -lcrypto"
-        LDFLAGS="$LDFLAGS -L$ssldir/lib"
         HAVE_OPENSSL=yes
         AC_DEFINE(HAVE_OPENSSL, 1, [Define if you have the OpenSSL.])
     fi
