--- freetds-0.95.8/configure.ac.orig	2015-06-22 11:00:49.000000000 +0200
+++ freetds-0.95.8/configure.ac	2015-06-25 21:14:01.183204149 +0200
@@ -568,12 +568,10 @@
 	if test -f "$with_unixodbc/bin/odbc_config"; then
 		ODBC_CONFIG="$with_unixodbc/bin/odbc_config"
 		ODBC_INC=`$ODBC_CONFIG --include-prefix`
-		ODBC_LDFLAGS="-L`$ODBC_CONFIG --lib-prefix`"
 	else
 		# if not available do not use system default
 		ODBC_CONFIG=""
 		ODBC_INC="$with_unixodbc/include"
-		ODBC_LDFLAGS="-L$with_unixodbc/lib"
 	fi
 	test -r "$ODBC_INC/sql.h" || AC_ERROR([sql.h not found])
 	test -r "$ODBC_INC/sqlext.h" || AC_ERROR([sqlext.h not found])
--- freetds-1.00.13/m4/check_openssl.m4.orig	2016-07-29 19:20:01.111369876 +0200
+++ freetds-1.00.13/m4/check_openssl.m4	2016-07-29 19:20:37.874702783 +0200
@@ -18,7 +18,6 @@
                 found_ssl="yes"
                 CFLAGS="$CFLAGS -I$ssldir/include"
                 NETWORK_LIBS="$NETWORK_LIBS -lssl -lcrypto"
-                LDFLAGS="$LDFLAGS -L$ssldir/lib"
                 break
             fi
         done
